addons/hr_timesheet/models/hr_timesheet.py

```diff
@@ -124,35 +124,78 @@ def _compute_department_id(self):
-
+        user_ids = []
+        employee_ids = []
+        # 1/ Collect the user_ids and employee_ids from each timesheet vals
         for vals in vals_list:
-            # when the name is not provide by the 'Add a line', we set a default one
-            if vals.get('project_id') and not vals.get('name'):
-                vals['name'] = '/'
             vals.update(self._timesheet_preprocess(vals))
+            if not vals.get('project_id'):
+                continue
+            if not vals.get('name'):
+                vals['name'] = '/'
+            employee_id = vals.get('employee_id')
+            user_id = vals.get('user_id', default_user_id)
+            if employee_id and employee_id not in employee_ids:
+                employee_ids.append(employee_id)
+            elif user_id not in user_ids:
+                user_ids.append(user_id)
 
         # Although this make a second loop on the vals, we need to wait the preprocess as it could change the company_id in the vals
         # TODO To be refactored in master
-        employees = self.env['hr.employee'].sudo().search([('user_id', 'in', user_ids)])
-        employee_for_user_company = defaultdict(dict)
+        # 2/ Search all employees related to user_ids and employee_ids, in the selected companies
+        Employee = self.env['hr.employee']
+        employees = Employee.sudo().with_context(active_test=False).search([
+            '&', '|', ('user_id', 'in', user_ids), ('id', 'in', employee_ids), ('company_id', 'in', self.env.companies.ids)
+        ]) or Employee
+
+        #                 ┌─── active/in companies ────────> was found with... ─── employee_id ───> (A) There is nothing to do, we will use this employee_id
+        # 3/ Each employee                                                    └─── user_id ───────> (B)** We'll need to select the right employee for this user
+        #                 └─── archived/not in companies ──> (C) We raise an error as we can't create a timesheet for an archived employee
+        # ** We can rely on the user to get the employee_id if
+        #    he has an active employee in the company of the timesheet
+        #    or he has only one active employee for all selected companies
+        active_employee_ids_in_vals = []
+        employee_per_company_per_user = defaultdict(dict)
         for employee in employees:
-            employee_for_user_company[employee.user_id.id][employee.company_id.id] = employee.id
-
-        employee_ids = set()
+            if employee.active:
+                if employee.id in employee_ids:
+                    active_employee_ids_in_vals.append(employee.id)
+                else:
+                    employee_per_company_per_user[employee.user_id.id][employee.company_id.id] = employee.id
+
+        # 4/ Put valid employee_id in each vals
+        error_msg = _('Timesheets must be created with an active employee in the selected companies.')
```

Comment:
To avoid doing a query to translate the error message when there is no error to raise we could do a lambda function to return the translated message. By doing that, we will only translate the error message when we have to raise the error message. 🙂 
```suggestion
        get_error_msg = lambda: _('Timesheets must be created with an active employee in the selected companies.')
```